src/templates/plugin.main.js

```diff
@@ -153,18 +153,22 @@ export default async (context) => {
     let redirectPath = ''
 
     const isStaticGenerate = process.static && process.server
+    const isDifferentLocale = getLocaleFromRoute(route) !== newLocale
+    const isPrefixAndDefaultStrategy = options.strategy === Constants.STRATEGIES.PREFIX_AND_DEFAULT
+    const isDefaultLocale = newLocale === options.defaultLocale
+
     // Decide whether we should redirect to a different route.
     if (
       !isStaticGenerate &&
       !app.i18n.differentDomains &&
       options.strategy !== Constants.STRATEGIES.NO_PREFIX &&
       // Skip if already on the new locale unless the strategy is "prefix_and_default" and this is the default
-      // locale, in which case we might still redirect as we prefer unprefixed route in this case.
-      (getLocaleFromRoute(route) !== newLocale || (options.strategy === Constants.STRATEGIES.PREFIX_AND_DEFAULT && newLocale === options.defaultLocale))
+      // locale, in which case we might still redirect as we prefer unprefixed route in this case, but let user a
+      // possibility to disable this behavior by switching disableDefaultRedirect option to true.
+      (isDifferentLocale || (isPrefixAndDefaultStrategy && isDefaultLocale))
     ) {
-      // The current route could be 404 in which case attempt to find matching route using the full path since
-      // "switchLocalePath" can only find routes if the current route exists.
-      const routePath = app.switchLocalePath(newLocale) || app.localePath(route.fullPath, newLocale)
```

Comment:
[rchl](https://github.com/rchl) With my current implementation switchLocalePath() always redirects me to the prefixed path and I couldn't think of anything better yet :(